lib/iex/lib/iex/history.ex

```diff
@@ -1,140 +1,124 @@
 
   # Checks val and each of its elements (if it is a list or a tuple)
-  # recursively to see if it has any binaries
+  # recursively to see if it has any large binaries (outside of the heap).
   defp has_binary(val) do
     try do
       has_bin(val)
     catch
-      :throw, true -> true
+      :throw, :found -> true
     end
   end
 
-  # Worker function used by has_binary. Throws when the first binary of the
-  # minimum specified size is found
-  defp has_bin(val) when is_tuple(val) do
-    has_bin(val, tuple_size(val)-1)
-  end
+  defp has_bin(val) when is_tuple(val),
+    do: has_bin(val, tuple_size(val) - 1)
 
-  defp has_bin([h|t]) do
-    has_bin(h)
-    has_bin(t)
+  defp has_bin([head | tail]) do
+    has_bin(head)
+    has_bin(tail)
   end
 
-  defp has_bin(val) when byte_size(val) > 64 do
-    throw true
-  end
+  defp has_bin(val) when byte_size(val) > 64,
+    do: throw(:found)
 
-  defp has_bin(_) do
-    false
-  end
+  defp has_bin(_), do: false
 
-  defp has_bin(_, -1) do
-    false
-  end
+  defp has_bin(_, -1), do: false
 
   defp has_bin(tuple, index) do
     has_bin(elem(tuple, index))
-    has_bin(tuple, index-1)
+    has_bin(tuple, index - 1)
   end
+end
 
-  # Based on https://github.com/erlang/otp/blob/7dcccee4371477e983f026db9e243cb66900b1ef/lib/stdlib/src/shell.erl#L1401
-  defp collect_garbage do
-    :erlang.garbage_collect(self())
-    collect_garbage Process.whereis(:user)
-    collect_garbage Process.group_leader()
-    :erlang.garbage_collect()
-  end
+defmodule IEx.History do
+  @moduledoc false
 
-  defp collect_garbage(process) do
-    try do
-      :erlang.garbage_collect(process)
-    catch
-      _, _ -> nil
-    end
-  end
+  alias IEx.History.State
 
   @doc """
-  Enumerates over all items in the history starting from the oldest one and
-  applies `fun` to each one in turn.
+  Starts IEx process variables. All history
+  information is kept in the process dictionary.
   """
-  def each(fun) do
-    each_pair(fn _, item -> fun.(item) end)
+  def start_link() do
+    Agent.start_link(fn -> State.new end)
   end
 
-  # Private helper that invokes fun with both key and item.
-  defp each_pair(fun) do
-    each_pair(Process.get(:iex_history_start_counter),
-              Process.get(:iex_history_counter),
-              fun)
+  @doc """
+  Appends one entry to the history.
+  """
+  def append(pid, entry, limit) do
+    should_collect = Agent.get_and_update(pid, fn entries ->
+      State.append(entries, entry) |> State.prune(limit)
+    end)
+    if should_collect do
+      collect_garbage(pid)
+    end
   end
 
-  defp each_pair(counter, max_counter, fun) when counter < max_counter do
-    key = {:iex_history, counter}
-    entry = Process.get(key)
-    fun.(key, entry)
-    each_pair(counter+1, max_counter, fun)
+  @doc """
+  Removes all entries from the history and forces a garbage collection cycle.
+  """
+  def reset(pid) do
+    Agent.update(pid, fn _ -> State.new end)
+    collect_garbage(pid)
```

Comment:
Yes, as I see it every process that has dealt with out of the heap binary have to be garbage collected.